hamt/partial-enum.md

```diff
@@ -0,0 +1,29 @@
+
+Note on terminology: "partial" may not be the most apt term here. It is just used as a contrast with the *full* enumeration of `ForEachLink` and `EnumLinksAsync`. The exact nature of the enumeration proposed needs to be further elucidated in the proposal here.
+
+*tl;dr*: keep fetched nodes from the enumeration in the `Shard` structure (as if they had been fetched as a consequence of an `AddChild`/`RemoveChild` call, which are retained throughout the lifetime of the structure/directory).
+
+## Trade memory for bandwidth
+
+The key insight of this proposal is the trade-off between memory (MEM) consumed by the `HAMTDirectory`/`Shard` and the bandwidth (BW) used to fetch the different nodes comprising the full `HAMTDirectory` DAG during enumeration.
+
+We want to minimize BW in the sense of not fetching *all* nodes at the same time but only as many needed to check if we are still above the transition threshold (`HAMTShardingSize`). It is not so much the *aggregated* BW we want to minimize (in the long run we may end up traversing the whole DAG) but the "burst" consumed in a single `Directory` API call when evaluating the transition (ideally we need only a number of nodes whose entries aggregated add up to `HAMTShardingSize`). This burst of enumeration then need not be full (all nodes) and hence we refer to partial enumeration. For this partial enumeration to be meaningful it needs to *not* overlap: we can't count the same node/entry twice toward the aggregated directory size. This means we need to track what was enumerated before: this is where we need to expense MEM.
+
+## HAMT/Shard structure
+
+The most meaningful way to organize this MEM is to leverage the `hamt.Shard` implementation which already stores the fetched nodes from the network (`loadChild`) in its internal `childer` structure. When we modify a shard/node we fetch it, modify it and keep it in memory (for future edits). Independent of the previous, when we enumerate the links of the `HAMTDirectory` we walk the DAG fetching the same nodes but discard them afterward (using them transiently as the path to follow to get to the leaves with the directory entries/links). (Both codes are flagged as part of this PR.)
+
+## Implementation
+
+The proposed implementation change is rather straightforward then: modify `makeAsyncTrieGetLinks` to store the fetched nodes (likely unifying some of its logic with `loadChild`).
+
+We can then create a low-level internal function that fetches only a part of the DAG (canceling the context after a certain number of nodes are loaded; we don't really care which). After that we enumerate only loaded nodes (so only a partial DAG, but enough to check against `HAMTDirectory`). The "partial"term of this proposal is more crucially applied not to the enumeration itself but to the fetching process: we only fetch part of the DAG and then enumerate those nodes and compute their size.
```

Comment:
> We can then create a low-level internal function that fetches only a part of the DAG

Is the order deterministic or non-deterministic? If non-deterministic then we'll end up doing more network requests than necessary since we won't utilize the cache of data we have on disk.

> canceling the context after a certain number of nodes are loaded; we don't really care which

nit: cancelling contexts is generally less good than just aborting your function as you have to wait for the context cancellation to get checked.

> After that we enumerate only loaded nodes

I'm not sure how annoying this is given the dagservice that you're handed.

Reply:
> Is the order deterministic or non-deterministic?

Non-deterministic, what comes faster, from the PR:
 
> fetches only a part of the DAG (canceling the context after a certain number of nodes are loaded; **we don't really care which**).

---------------------------------------------------

> If non-deterministic then we'll end up doing more network requests than necessary since we won't utilize the cache of data we have on disk.

We are implicitly caching in memory by storing the retrieved `Shard`s instead of discarding them as the cited function does. The cache is the structure itself (we can move it to the disk if your assumption above holds, but you'd still need to keep track of which shards you already have on disk, otherwise I'm missing something in your proposal,  if so please correct me).

-----------------------------------

> I'm not sure how annoying this is given the dagservice that you're handed.

They are already in memory in the `childer` structure, you wouldn't go through the DAG service.